src/game.cpp

```diff
@@ -12300,6 +12310,61 @@ void game::wield(int pos)
     }
 }
 
+void game::open_bag(int pos)
+{
+    item &bag = u.i_at(pos);
+    if(u.rate_action_open_bag(&bag) != HINT_GOOD) { return; }
+    //[davek] TODO: make appropriate overload for 'key' in uimenu_cb for handling
+    uimenu bmenu;
+    //[davek] FIXME: make this return to inv instead of menu
+    bmenu.return_invalid = true;
+    bmenu.text = string_format(_("You stare at the %s"), bag.tname(1, false).c_str());
+    bmenu.addentry(1,  !bag.is_container_full(),  's', _("Store an item"));
+    bmenu.addentry(2,  !bag.is_container_empty(), 't', _("Take an item"));
+    bmenu.query();
+    int decision = bmenu.ret;
+    bmenu.reset();
+    bmenu.return_invalid = true;
+
+    switch(decision) {
+/*** Storing an item ***/
+    case 1: {
+        bmenu.text = string_format(_("You ponder what to put in the %s..."), bag.tname(1, false).c_str());
+        int i = 0;
+        // used for getting an index back to move around the item in a sensible way
+        std::map<int, item *> ur_junk;
+        auto const &pizza = u.inv.slice();
+        for(auto &elem : pizza) {
+            item &it = elem->front();
+            ur_junk[i] = &it;
+            bmenu.addentry(++i, true, -1, elem->front().tname());
+        }
+        bmenu.query();
+        // invalid return value, requery!
+        if(!is_between(1, bmenu.ret, i, true)) { return; }
+        bag.has_been_opened = true;
+        //[davek] TODO: explain volume_factor
+        u.store(&bag, ur_junk[bmenu.ret-1], "null", bag.storage_used());
+        break; } // end of case 1
+
+/*** Taking an item ***/
+    case 2: {
+        bmenu.text = string_format(_("You open the %s, and see..."), bag.tname(1, false).c_str());
+        for(unsigned int i = 0; i < bag.contents.size(); ++i) {
+            bmenu.addentry((i + 1), true, -1, bag.contents[i].tname());
+        }
+        bmenu.query();
+        int index = clamp_value(0, (bmenu.ret - 1), bag.contents.size());
```

Comment:
I know that this is the cause of removing an item even if you don't select it [via escape to quit], but I put this here to prevent OOB errors with vector for the time being.